lib/posix/spawn/child.rb

```diff
@@ -244,14 +257,27 @@ def read_and_write(input, stdin, stdout, stderr, timeout=nil, max=nil)
 
           # read from stdout and stderr streams
           ready[0].each do |fd|
-            buf = (fd == stdout) ? @out : @err
+            chunk = nil
             begin
-              buf << fd.readpartial(BUFSIZE)
+              chunk = fd.readpartial(BUFSIZE)
```

Comment:
I could be wrong here (cc @tenderlove) but I'm pretty sure the previous code would actually create a new string (down inside `readpartial`), then _that_ string would have been appended to `buf`. Which would require potentially resizing that buffer first, then the `memcpy`.

This new code just keeps that first string as a local var first, so we can later determine where to write it. In the default case we're using a `StringIO` so the result is essentially the same as before (potential buffer resize then a copy). Though iirc we saw some pretty significant speedups by using an array to keep track of chunks, then calling `join` at the end when it was all needed. The reason for that is because it avoids the reallocation of the resulting buffer as we're reading, and instead allows `join` to allocate a buffer exactly the size that's needed then copying all the chunks in to it.

Basically this (the old way):

``` ruby
buffer = ""
buffer << "one,"
buffer << "two,"
buffer << "three"

return buffer
```

vs this (the array optimized version I just mentioned):

``` ruby
buffer = []
buffer << "one,"
buffer << "two,"
buffer << "three"

# buffer is an array with 3 elements at this point
# and this join call figures out how big all of the strings inside are, then creates a single buffer to append them to.
return buffer.join
```

Using that approach efficiently may change the API contract here slightly though...